header.go

```diff
@@ -6,12 +6,12 @@ import (
 )
 
 // Header abstracts all methods required to perform header sync.
-type Header interface {
+type Header[H any] interface { // TODO: Ideally, this should be Header[H Header[H]], but GO	does not support recursive type definitions.
```

Comment:
I tried `Header[H Header[H]]` during initial design prototyping, and it didn't work as TODO says, but it turned out there is another way to achieve similar parametrization through `any` or `interface`.

Technically, it allows the header to be paramed with any type, BUT: the services and other consumers of that type can do `[H Header[H]]` ensuring type safety, like `Exchange[H Header[H]]`